### PR TITLE
♻️ Base foreground computation on page lifecycle states

### PR DESCRIPTION
**Pull request by amortemousque**
## Motivation

Follow up of https://github.com/DataDog/browser-sdk/pull/2229: Base foreground computation on page lifecycle states

## Changes

- Add `pageStateHistory.isActiveAt()` to replace `foregroundContexts.isInForegroundAt()`
- Base foreground computation on `pageStateHistory` (implement `mapToForegroundPeriods()`)

## Testing

- [x] Local
- [ ] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/rum-core/src/domain/contexts/pageStateHistory.ts [hunk 1]
```diff
@@ -98,6 +99,10 @@ export function startPageStateHistory(
 
       return pageStateServerEntries
     },
+    isActiveAt: (startTime: RelativeTime) => {
```
Comment (bcaudan) [inline packages/rum-core/src/domain/contexts/pageStateHistory.ts:102]: **💬 suggestion:** ‏more explicit naming
```suggestion
    isInActivePageStateAt: (startTime: RelativeTime) => {
```

File: packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts [hunk 2]
```diff
@@ -98,9 +98,9 @@ function processAction(
     },
     autoActionProperties
   )
-  const inForeground = foregroundContexts.isInForegroundAt(action.startClocks.relative)
-  if (inForeground !== undefined) {
-    actionEvent.view = { in_foreground: inForeground }
+  const inActivePageState = pageStateHistory.isActiveAt(action.startClocks.relative)
+  if (inActivePageState !== undefined) {
+    actionEvent.view = { in_foreground: inActivePageState }
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts:103]: **❗ warning:** ‏ ‏FMU, `inActivePageState` can't be `undefined`, there is maybe a missing test here

File: packages/rum-core/src/domain/contexts/foregroundContexts.ts [hunk 3]
```diff
@@ -1,129 +1,32 @@
...
-    const foregroundPeriod = foregroundPeriods[i]
-    if (foregroundPeriod.end !== undefined && startTime > foregroundPeriod.end) {
-      break
-    }
-    if (
-      startTime > foregroundPeriod.start &&
-      (foregroundPeriod.end === undefined || startTime < foregroundPeriod.end)
-    ) {
-      return true
+// Todo: Remove in the next major release
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/rum-core/src/domain/contexts/foregroundContexts.ts:11]: We could have some customers that rely on before_send and expect resource.duration to be defined

File: packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts (hunk 2 above)
Comment (amortemousque) [reply packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts:103]: Indeed! I checked and this comes from when there was a feature flag. I'll remove this check

**Review by BenoitZugmeyer — APPROVED** [review]
Great work!

File: packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts [hunk 4]
```diff
@@ -95,13 +95,11 @@ function processAction(
...
       type: RumEventType.ACTION as const,
+      view: { in_foreground: pageStateHistory.isInActivePageStateAt(action.startClocks.relative) },
     },
     autoActionProperties
   )
-  const inForeground = foregroundContexts.isInForegroundAt(action.startClocks.relative)
-  if (inForeground !== undefined) {
-    actionEvent.view = { in_foreground: inForeground }
-  }
+
```
Comment (bcaudan) [inline packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts:102]: **💬 suggestion:** ‏what about keeping the same logic of not adding the `view.in_foreground` if there is no active page state?

**Review by bcaudan — APPROVED** [review]

File: packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts (hunk 4 above)
Comment (amortemousque) [reply packages/rum-core/src/domain/rumEventsCollection/action/actionCollection.ts:102]: The previous `foregroundContexts.isInForegroundAt(action.startClocks.relative)` always return a boolean so `inForeground !== undefined` is always true

**Pull request merged**